src/infra/core/src/ingest_service_impl.rs

```diff
@@ -909,85 +927,120 @@ impl IngestServiceImpl {
             maybe_multi_listener.unwrap_or_else(|| Arc::new(NullIngestMultiListener));
 
-        tracing::info!(?requests, "Ingesting multiple datasets");
-
-        let futures: Vec<_> = requests
+        let futures: Vec<_> = dataset_refs
             .iter()
-            .map(|req| {
-                self.do_ingest(
-                    &req.dataset_ref,
-                    options.clone(),
-                    req.fetch_override.clone(),
-                    |hdl| multi_listener.begin_ingest(hdl),
-                )
+            .map(|dataset_ref| {
+                self.do_ingest(dataset_ref, options.clone(), None, |hdl| {
+                    multi_listener.begin_ingest(hdl)
+                })
             })
             .collect();
 
         let results = futures::future::join_all(futures).await;
-        requests
+        dataset_refs
             .into_iter()
-            .map(|r| r.dataset_ref)
             .zip(results)
+            .map(|(dataset_ref, result)| IngestResponse {
+                dataset_ref,
+                result,
+            })
             .collect()
     }
+
+    #[tracing::instrument(level = "info", skip_all, fields(%dataset_ref, %data_url))]
+    async fn push_ingest_from_url(
+        &self,
+        dataset_ref: &DatasetRef,
+        data_url: url::Url,
+        listener: Option<Arc<dyn IngestListener>>,
+    ) -> Result<IngestResult, IngestError> {
+        let fetch = FetchStep::Url(FetchStepUrl {
+            url: data_url.to_string(),
+            event_time: Some(EventTimeSource::FromSystemTime),
+            cache: None,
+            headers: None,
+        });
+
+        self.do_ingest(
+            dataset_ref,
+            PollingIngestOptions {
+                fetch_uncacheable: true,
+                exhaust_sources: false,
+            },
+            Some(fetch),
+            |_| listener,
+        )
+        .await
+    }
+
+    #[tracing::instrument(level = "info", skip_all, fields(%dataset_ref))]
+    async fn push_ingest_from_stream(
+        &self,
+        dataset_ref: &DatasetRef,
+        mut data: Box<dyn AsyncRead + Send + Unpin>,
+        listener: Option<Arc<dyn IngestListener>>,
+    ) -> Result<IngestResult, IngestError> {
+        use tokio::io::{AsyncReadExt, AsyncWriteExt};
+
+        // Save stream to a file in cache
+        //
+        // TODO: Breaking all architecture layers here - need to extract cache into a
+        // service
+        let path = self
+            .cache_dir
+            .join(self.get_random_cache_key("push-ingest-"));
+
+        {
+            let mut file = tokio::fs::File::create(&path).await.int_err()?;
+            let mut buf = [0u8; 2048];
+            loop {
+                let read = data.read(&mut buf).await.int_err()?;
+                if read == 0 {
+                    break;
+                }
+                file.write_all(&buf[..read]).await.int_err()?;
+            }
+            file.flush().await.int_err()?;
+        }
```

Comment:
Yes, this thing looks extremely bad

Reply:
We must set a size limit and throw an error in the API handler, if it is exceeded.
Right now this is an easy attack vector against our server - I don't think that container has more than a few gigs of free space in the cluster.